pkg/azclient/factory.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azclient
+
+import (
+	"strings"
+	"sync"
+
+	"sigs.k8s.io/cloud-provider-azure/pkg/azclient/diskclient"
+	"sigs.k8s.io/cloud-provider-azure/pkg/azclient/policy/ratelimit"
+)
+
+type ClientFactory struct {
+	*ClientFactoryConfig
+	ResourceManagerEndpoint string
+	diskClientRegistry      sync.Map
+	*AuthProvider
+}
+
+func NewClientFactory(config *ClientFactoryConfig, resourceManagerEndpoint string, authProvider *AuthProvider) *ClientFactory {
+	return &ClientFactory{
+		ClientFactoryConfig:     config,
+		ResourceManagerEndpoint: resourceManagerEndpoint,
+		AuthProvider:            authProvider,
+	}
+}
+
+func (factory *ClientFactory) GetDiskClient(subscription string) (diskclient.Interface, error) {
+	subID := strings.ToLower(subscription)
+
+	options := factory.GetDefaultResourceClientOption()
+	//add ratelimit policy
+	ratelimitOption := factory.ClientFactoryConfig.GetRateLimitConfig("diskRateLimit")
+	rateLimitPolicy := ratelimit.NewRateLimitPolicy(ratelimitOption)
+	options.ClientOptions.PerCallPolicies = append(options.ClientOptions.PerCallPolicies, rateLimitPolicy)
+
+	cred, err := factory.AuthProvider.GetAzIdentity()
+	if err != nil {
+		return nil, err
+	}
+	newClient, err := diskclient.New(subscription, cred, options)
+	if err != nil {
+		return nil, err
+	}
+	client, _ := factory.diskClientRegistry.LoadOrStore(subID, &newClient)
```

Comment:
In what scenarios could `diskClientRegistry` be used, apart from this one?

Reply:
It is the sole reference it has.

Reply:
I fail to see the reason why we need `diskClientRegistry`. If the purpose is to reduce the `new client` calls, it doesn't seem to help.

Reply:
throttling/ratelimit policies are appended to the client. if we reuse the client, all of timers/ counters will not be reset.

Reply:
I'm not quite following. Are you trying to reuse the client or not? If yes, could you please load and return the existing client if it has already been created? And it would help reduce the `new client` calls.

```go

client, found := registry.Load(...)
if found {
  return client
}
client = newClient(...)
registry.LoadOrStore(client)
...
```

Reply:
It is not thread-safe.

Reply:
the client returned in the last line may not the one we put in the params.